plugin/builtin/buildbaron/build_baron.go

```diff
@@ -156,16 +214,155 @@ func (bbp *BuildBaronPlugin) buildFailuresSearch(w http.ResponseWriter, r *http.
+	jql := taskToJQL(t, js.bbProj.TicketSearchProjects)
+
+	results, err := js.jiraHandler.JQLSearch(jql, 0, -1)
+	if err != nil {
+		return nil, err
+	}
+
+	return results.Issues, nil
+}
+
+func (js *jiraSuggest) GetTimeout() time.Duration {
+	return time.Duration(0)
+}
+
+type altEndpointSuggest struct {
+	bbProj bbProject
+}
+
+// parseResponse converts the Build Baron tool's suggestion response into JIRA ticket results.
+func (aes *altEndpointSuggest) parseResponse(r io.ReadCloser) ([]thirdparty.JiraTicket, error) {
+	data := struct {
+		Status      string `json:"status"`
+		Suggestions []struct {
+			TestName string `json:"test_name"`
+			Issues   []struct {
+				Key         string `json:"key"`
+				Summary     string `json:"summary"`
+				Status      string `json:"status"`
+				Resolution  string `json:"resolution"`
+				CreatedDate string `json:"created_date"`
+				UpdatedDate string `json:"updated_date"`
+			}
+		} `json:"suggestions"`
+	}{}
+
+	if err := util.ReadJSONInto(r, &data); err != nil {
+		return nil, err
+	}
+
+	if data.Status != "ok" {
+		return nil, errors.Errorf("Build Baron suggestions weren't ready: status=%s", data.Status)
+	}
+
+	var tickets []thirdparty.JiraTicket
+	for _, suggestion := range data.Suggestions {
+		for _, issue := range suggestion.Issues {
+			ticket := thirdparty.JiraTicket{
+				Key: issue.Key,
+				Fields: &thirdparty.TicketFields{
+					Summary: issue.Summary,
+					Created: issue.CreatedDate,
+					Updated: issue.UpdatedDate,
+					Status:  &thirdparty.JiraStatus{Name: issue.Status},
+				},
+			}
+
+			if issue.Resolution != "" {
+				ticket.Fields.Resolution = &thirdparty.JiraResolution{Name: issue.Resolution}
+			}
+
+			tickets = append(tickets, ticket)
+		}
+	}
+
+	if len(tickets) == 0 {
+		// We treat not having suggestions as an error so that it causes fallback to occur in a
+		// unified way.
+		return nil, errors.New("no suggestions found")
+	}
+
+	return tickets, nil
+}
+
+func (aes *altEndpointSuggest) Suggest(ctx context.Context, t *task.Task) ([]thirdparty.JiraTicket, error) {
+	client := util.GetHTTPClient()
+	defer util.PutHTTPClient(client)
+
+	path := aes.bbProj.AlternativeEndpointURL.Path
+	path = strings.Replace(path, "{task_id}", t.Id, -1)
+	path = strings.Replace(path, "{execution}", strconv.Itoa(t.Execution), -1)
+
+	altEndpointURL := url.URL{
+		Scheme: aes.bbProj.AlternativeEndpointURL.Scheme,
+		Host:   aes.bbProj.AlternativeEndpointURL.Host,
+		Path:   path,
+	}
+	req, err := http.NewRequest(http.MethodGet, altEndpointURL.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	creds := aes.bbProj.AlternativeEndpointCredentials
+	if creds.Username != "" {
+		req.SetBasicAuth(creds.Username, creds.Password)
+	}
+
+	req = req.WithContext(ctx)
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode >= 300 || resp.StatusCode < 200 {
+		defer resp.Body.Close()
```

Comment:
This defer must be moved immediately below client.Do(...) in a conditional:
```
if resp != nil {
        defer resp.Body.Close();
}
```

There are scenarios where `resp != nil && err != nil`

Reply:
The pattern that we use for request handling is roughly the same here as in the agent. We don't close in cases where err is non-nil, and it does look like we routinely close the body more than once in some situations.